src/python/pants/core/util_rules/gen_rule.py

```diff
@@ -0,0 +1,145 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import os
+
+from pants.core.target_types import (
+    FilesSources,
+    GenRuleCommandField,
+    GenRuleOutputsField,
+    GenRuleSources,
+    GenRuleToolsField,
+)
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.fs import (
+    EMPTY_SNAPSHOT,
+    AddPrefix,
+    CreateDigest,
+    Digest,
+    Directory,
+    GlobMatchErrorBehavior,
+    MergeDigests,
+    PathGlobs,
+    Snapshot,
+)
+from pants.engine.process import (
+    BashBinary,
+    BinaryNotFoundError,
+    BinaryPathRequest,
+    BinaryPaths,
+    Process,
+    ProcessResult,
+    SearchPath,
+)
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import (
+    GeneratedSources,
+    GenerateSourcesRequest,
+    Sources,
+    TransitiveTargets,
+    TransitiveTargetsRequest,
+)
+from pants.engine.unions import UnionRule
+from pants.util.logging import LogLevel
+
+
+class GenerateFilesFromGenRuleRequest(GenerateSourcesRequest):
+    input = GenRuleSources
+    output = FilesSources
+
+
+@rule(desc="Running gen_rule", level=LogLevel.DEBUG)
+async def run_gen_rule(
+    request: GenerateFilesFromGenRuleRequest, bash: BashBinary
+) -> GeneratedSources:
+    gen_rule = request.protocol_target
+    working_directory = gen_rule.address.spec_path
+    command = gen_rule[GenRuleCommandField].value
+    tools = gen_rule[GenRuleToolsField].value
+    outputs = gen_rule[GenRuleOutputsField].value
+    source_globs = gen_rule[GenRuleSources].value or ()
+
+    if not (command and tools and outputs):
```

Comment:
The docstring says that the command runs "for side effects": the purest definition of a "side effect" in my mind is an effect which is not an output, and instead mutates the world in some way. If that's really what you meant by side effect, then allowing commands without `outputs` to run would make sense.

But also, it seems like raising a `ValueError` for a noop target would make sense. You can do it directly on the `Target` definition similarly to #12925.

Reply:
Yeah, not output should def be allowed. There was an issue with running the Process though, that caught me a bit off guard, that had to do with digests (input or output), but don't recall the details right now. Will find out again if it was the output one..